### PR TITLE
Rate limit event driven

### PR DESCRIPTION
**Pull request by RodrigoAD**

### CODE REVIEW

File: bootstrap/src/index.ts [hunk 1]
```diff
@@ -35,6 +37,21 @@ const withStatusCode: Middleware = async (execute) => async (input) => {
   return { ...rest, statusCode, data }
 }
 
+export const withRateLimit: Middleware = async (execute) => async (input) => {
+  const totalReqPerMin = rateLimit.getTroughput(rateLimit.Interval.MINUTE)
+  const participantReqPerMin = rateLimit.getTroughput(rateLimit.Interval.MINUTE, input)
+  const cost = rateLimit.getParticipantCost(participantReqPerMin)
+  const maxReqPerMin = rateLimit.getMaxReqAllowed(
+    totalReqPerMin.length + 1,
+    participantReqPerMin.length + 1,
```
Comment (RodrigoAD) [inline bootstrap/src/index.ts:46]: If the request succeed, we could assume that the request will be 1 more than the current one.

File: bootstrap/src/lib/cache-warmer/reducer.ts [hunk 2]
```diff
@@ -92,4 +92,3 @@ export const rootReducer = combineReducers({
   subscriptions: subscriptionsReducer,
   warmups: warmupReducer,
 })
-export type RootState = ReturnType<typeof rootReducer>
```
Comment (krebernisak) [inline bootstrap/src/lib/cache-warmer/reducer.ts:96]: We need this, the cache warmer reducer needs to export its state type.

File: bootstrap/src/lib/cache-warmer/side-effects.ts [hunk 3]
```diff
@@ -21,7 +21,7 @@ import {
   warmupUnsubscribed,
 } from './actions'
 import { WARMUP_REQUEST_ID, Config, get } from './config'
-import { RootState } from './reducer'
+import { RootState } from '../../index'
```
Comment (krebernisak) [inline bootstrap/src/lib/cache-warmer/side-effects.ts:24]: Please revert changes to this file.

We need to be aware of what depends on what. This warmer component needs to be contained, it should not depend on the application, but the application depends on the warmer component.

Comment (RodrigoAD) [reply bootstrap/src/lib/cache-warmer/side-effects.ts:24]: Yep, I know, but this was breaking the cache warmer. We need to tell somehow to `createEpicMiddleware` that Root State now has rate-limit as well, as both stores and reducers are combined.

File: bootstrap/src/lib/cache/index.ts [hunk 4]
```diff
@@ -150,8 +150,6 @@ export const withCache: Middleware = async (execute, options = defaultOptions())
 
     if (entry) {
       if (maxAge >= 0) {
-        logger.debug(`Cache: GET ${key}`, entry)
```
Comment (krebernisak) [inline bootstrap/src/lib/cache/index.ts:155]: Why did you remove this? I think we'd need this feature to update the already cached item with the new TTL.

If you don't want for warmer to recache the item, then don't set a `maxAge` to the input and that will achieve the same thing.

Comment (RodrigoAD) [reply bootstrap/src/lib/cache/index.ts:155]: This is keeping the cache values forever. As maxAge from the input is constantly changing, it updates the cache entry with the same values, but with extended maxAge

Comment (RodrigoAD) [reply bootstrap/src/lib/cache/index.ts:155]: I've included the max age coming from the RateLimit middleware into `data: { rateLimitMaxAge: ... }`. If the request contains `maxAge` would override the default value

File: bootstrap/src/lib/metrics/index.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+const collectDefaultMetrics = client.collectDefaultMetrics
+const Registry = client.Registry
+export const register = new Registry()
+
+collectDefaultMetrics({ register })
+
+export const httpRequestsTotal = new client.Counter({
+  name: 'http_requests_total',
+  help: 'The number of http requests this external adapter has serviced for its entire uptime',
+  labelNames: ['method', 'statusCode', 'apiKey', 'retry'] as const,
```
Comment (HenryNguyen5) [inline bootstrap/src/lib/metrics/index.ts:12]: I would merge `cacheWarmerRequests`, `httpRequestsCacheHits`, `httpRequestsDataProviderHits` into this metric as labels, like the following:
```ts
labelNames: ['method', 'statusCode', 'apiKey', 'retry', 'isCacheHit', 'isWarmerRequest', 'isDataProviderHit'] as const,
```

Comment (HenryNguyen5) [reply bootstrap/src/lib/metrics/index.ts:12]: This way, you can set those additional labels as boolean values

**Review by HenryNguyen5 — CHANGES_REQUESTED** [review]

File: bootstrap/src/lib/metrics/index.ts (hunk 5 above)
Comment (HenryNguyen5) [reply bootstrap/src/lib/metrics/index.ts:12]: Now that i think about it, we dont even need `isDataProviderHit` since we can infer that from `isCacheHit`

Comment (RodrigoAD) [reply bootstrap/src/lib/metrics/index.ts:12]: Perfect, didn't know that

**Comment by RodrigoAD** [comment]
Base branch got so many changes that this PR became unmaintainable. 
- Most of the functionality from this PR got merged here: #366 
- Closing in favor of #373
